src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -7065,12 +7065,49 @@ bool canFollowNullableType()
+                        // These contexts might be a type that is at the end of an expression.
+                        // For backward compatibility we want to consider a `?` token as part of the `?:` operator if possible.
+                        // However, if current token is an identifier, it might be beneficial to allow `?`
+                        // and treat this identifier as a designation. Nullable types in patterns are semantically invalid,
+                        // so we will get a nice error about that during binding.
+                        if (this.CurrentToken.Kind == SyntaxKind.IdentifierToken)
+                        {
+                            // Given that we are deciding whether we take `?:` operator or not while looking at state after `?`,
+                            // our focus is on `:` token as well. But `:` have special meaning in context of switch statements
+                            // and expressions (since `:` in switch expressions is recovered to `=>`). So consider the following case:
+                            // `case T? t:`. On its own `T ? t :` can be treated as a conditional expression with condition `T`,
+                            // 'whenTrue' `t` and 'whenFalse' not yet parsed. But it is way better to take `:` as case label end and
+                            // thus parse label condition as a declaration pattern `T? t`. `whenIsKeyword` flag can conveniently tell
+                            // whether we are in a 'top-level' state of a label, so if condition get parenthesized like `case (a ? b : c):`
+                            // we don't take the first `:` as label end and parse label condition as a parenthesized conditional expression
+                            if ((_termState.HasFlag(TerminatorState.IsExpressionOrPatternInCaseLabelOfSwitchStatement) ||
+                                 _termState.HasFlag(TerminatorState.IsPatternInSwitchExpressionArm)) && whenIsKeyword)
+                            {
+                                return true;
+                            }
+
+                            using var _ = GetDisposableResetPoint(resetOnDispose: true);
+                            this.EatToken();
+
+                            // If token after identifier starts an expression it is probably an error case, e.g. missing a `,` in a pattern
+                            if (CanStartExpression())
+                            {
+                                return true;
+                            }
+
+                            // These token either 100% end a pattern or start a new one (again, in cases with missing `,` and so on).
+                            // Note, that some cases of 'token starts a new pattern' are handled by condition above,
+                            // e.g. starting of type patterns
+                            return this.CurrentToken.Kind is SyntaxKind.OpenParenToken or SyntaxKind.CloseParenToken
```

Comment:
> These token either 100% end a pattern or start a new one

i don't see why.  imagine `x ? a()`

We see the identifier `a`, we then also see that `()` "Can Start Expression".  And then the current token's kind is an open paren.  But this def doesn't necessarily start a pattern.  it can just be part of a conditional like `x ? a() : ...`

Reply:
Yeah, I see my assumption broken here. Same thing with `x ? a[]` actually. Another question os how it is better to handle `or` and `and` in this case. On one hand `x is Type? and (1, 2)` is better to be parsed as pattern, while `x is type ? and(1, 2) : 0` is better to be parsed as conditional expression (since it is technically possible to have method/delegate with name `and` in the scope)

Reply:
Fixed. Unfortunatelly, 2 "good" cases were "broken" in the process

Reply:
i think it's fine to limit to a very narrow set of cases where we go off the rails and could do better.  For everything else, fall back to the existing behavior.